### ISSUE TITLE
SelectOneMenu: Accessibility enhancement render grouped with <optgroup> tags

### ISSUE THREAD

**Issue by nichirsch12** [primefaces/primefaces#12526]
### Description

I see that the "Grouping" dropdown on the [SelectOneMenu page](https://www.primefaces.org/showcase/ui/input/oneMenu.xhtml?jfwid=33b32) in the Primefaces showcase ends up being rendered as a `<ul>` with `<li>` elements. The group label items are not selectable but have a special "ui-selectonemenu-item-group" class. 

The problem I have is, when reading these with a screenreader, such as NVDA, it still includes these groupings in the count of items, and does not reference the group at all. Despite the fact that there are only 10 valid options to select from, arrowing through the list yields the following reading:

"Select One 1 of 12
Germany 3 of 12
...
Mexico 12 of 12"

### Describe the solution you would like

Ideally, this type of dropdown could work like the example listed [here in MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/optgroup), using the `<optgroup label="">` tag.

When using NVDA on this page, it correctly identifies that there are 6 selectable options available, and reads the groups as well. Arrowing through _this_ list yields the following reading, which is the behavior I would expect:

"Theropods grouping
Tyrannosaurus 1 of 6
Velociraptor 2 of 6
...
Sauropods grouping
Diplodocus 4 of 6"

**Comment by melloware** [comment]
Unfortunately to allow us the custom styling you see for things like Groups and Advanced they have been `<li>` that hide the real options.

![image](https://github.com/user-attachments/assets/2f5b6de2-bce9-4416-ab39-63bb9cbee829)

although i wonder if `role="optgroup"` for this list items is possible.

**Comment by melloware** [comment]
So here is a real ARIA example using `<li>` they use `role="presentation"` for those groups: https://www.w3.org/WAI/ARIA/apg/patterns/listbox/examples/listbox-grouped/

**Comment by nichirsch12** [comment]
Interesting, so then on that page each group is its own `<ul>` contained within a `<div>`. Definitely screenreads correctly.

**Comment by melloware** [comment]
@nichirsch12 can you try this Monkeypatch?  Just add this JS to a file that loads after PrimeFaces.  Then make sure to CTRL+F5 in your browser to force JS reload.

```js
if (PrimeFaces.widget.SelectOneMenu) {
  /**
   * Renders panel content based on hidden select.
   * @private
   * @param {boolean} initContentsAndBindItemEvents `true` to call {@link initContents} and {@link bindItemEvents}
   * after rendering, `false` otherwise.
   */
  PrimeFaces.widget.SelectOneMenu.prototype.renderPanelContentFromHiddenSelect = function(initContentsAndBindItemEvents) {
    if (this.cfg.renderPanelContentOnClient && this.itemsWrapper.children().length === 0) {
      var panelContent = '<div id="' + this.id + '_items" class="ui-selectonemenu-items ui-widget-content ui-widget ui-corner-all ui-helper-reset" role="listbox">';
      panelContent += this.renderSelectItems(this.input);
      panelContent += '</div>';

      this.itemsWrapper.append(panelContent);

      if (initContentsAndBindItemEvents) {
        this.initContents();
        this.bindItemEvents();
      }
    }
  };

  /**
   * Renders panel HTML-code for all select items.
   * @private
   * @param {JQuery} parentItem A parent item (select, optgroup) for which to render HTML code.
   * @param {boolean} [isGrouped] Indicated whether the elements of the parent item should be marked as grouped.
   * @return {string} The rendered HTML string.
   */
  PrimeFaces.widget.SelectOneMenu.prototype.renderSelectItems = function(parentItem, isGrouped) {
    var $this = this;
    var content = "";
    isGrouped = isGrouped || false;

    var opts = parentItem.children("option, optgroup");
    var hasOptgroup = opts.filter("optgroup").length > 0;

    if (!hasOptgroup && !isGrouped) {
      content += '<ul role="group" class="ui-selectonemenu-list ui-widget-content ui-widget ui-corner-all ui-helper-reset">';
    }

    content += opts.map(function(index, element) {
      return $this.renderSelectItem(element, isGrouped);
    }).get().join('');

    if (!hasOptgroup && !isGrouped) {
      content += '</ul>';
    }

    return content;
  };

  /**
   * Renders panel HTML code for one select item (group).
   * @private
   * @param {JQuery} item An option (group) for which to render HTML code.
   * @param {boolean} [isGrouped] Indicates whether the item is part of a group.
   * @return {string} The rendered HTML string.
   */
  PrimeFaces.widget.SelectOneMenu.prototype.renderSelectItem = function(item, isGrouped) {
    var $item = $(item);
    var isOptgroup = item.tagName === "OPTGROUP";
    var id = PrimeFaces.uuid();
    var escape = $item.data("escape");
    var label = isOptgroup ? $item.attr("label") : $item.text();

    if (escape) {
      label = isOptgroup ? $("<div>").text(label).html() : $item.html();
      label = label === "&nbsp;" ? " " : label;
    }

    var cssClass = isOptgroup ? "ui-selectonemenu-item-group ui-corner-all" :
      "ui-selectonemenu-item ui-selectonemenu-list-item ui-corner-all" +
      (isGrouped ? " ui-selectonemenu-item-group-children" : "") +
      ($item.data("noselection-option") ? " ui-noselection-option" : "");

    var content = isOptgroup ? '<ul role="group" class="ui-selectonemenu-list ui-widget-content ui-widget ui-corner-all ui-helper-reset" aria-labelledby="' + id + '">' : '';

    content += '<li id="' + id + '" class="' + cssClass + '" tabindex="-1" role="' + (isOptgroup ? "presentation" : "option") + '"';

    var title = $item.data("title");
    if (title) {
      content += ' title="' + PrimeFaces.escapeHTML(title) + '"';
    }
    if ($item.is(':disabled')) {
      content += ' disabled';
    }

    var dataLabel = escape ? label.replaceAll('"', '&quot;') : PrimeFaces.escapeHTML(label, true);
    content += ' data-label="' + dataLabel + '">' + label + '</li>';

    if (isOptgroup) {
      content += this.renderSelectItems($item, true) + '</ul>';
    }

    return content;
  };

};
```

**Comment by melloware** [comment]
I have deployed this patch here if you want to test it: https://quarkus-faces-melloware-8a6a34c1.koyeb.app/ui/input/oneMenu.xhtml

**Comment by nichirsch12** [comment]
Seems like it is part of the way there - It does read the grouping, and it counts the correct number of items. But now the keyboard navigation seems broken. If I hit spacebar to open the dropdown, I cannot use the arrow keys to get to a different group, I can only go up and down within a group.

**Comment by melloware** [comment]
Good feedback that makes sense let me look at the keyboard support.

**Comment by melloware** [comment]
@nichirsch12 can you try again: https://quarkus-faces-melloware-8a6a34c1.koyeb.app/ui/input/oneMenu.xhtml

Keyboard i think is fixed.

**Comment by nichirsch12** [comment]
Yeah that all seems to work perfectly! Thanks for giving this your time.

**Comment by melloware** [comment]
Ok I will provide a proper MonkeyPatch so you can start using it right now in your app

**Comment by melloware** [comment]
MonkeyPatch so you can use this right now in your code.

```js
if (PrimeFaces.widget.SelectOneMenu) {
  /**
   * Renders panel content based on hidden select.
   * @private
   * @param {boolean} initContentsAndBindItemEvents `true` to call {@link initContents} and {@link bindItemEvents}
   * after rendering, `false` otherwise.
   */
  PrimeFaces.widget.SelectOneMenu.prototype.renderPanelContentFromHiddenSelect = function(initContentsAndBindItemEvents) {
    if (this.cfg.renderPanelContentOnClient && this.itemsWrapper.children().length === 0) {
      var panelContent = '<div id="' + this.id + '_items" class="ui-selectonemenu-items ui-widget-content ui-widget ui-corner-all ui-helper-reset" role="listbox">';
      panelContent += this.renderSelectItems(this.input);
      panelContent += '</div>';

      this.itemsWrapper.append(panelContent);

      if (initContentsAndBindItemEvents) {
        this.initContents();
        this.bindItemEvents();
      }
    }
  };

  /**
   * Renders panel HTML-code for all select items.
   * @private
   * @param {JQuery} parentItem A parent item (select, optgroup) for which to render HTML code.
   * @param {boolean} [isGrouped] Indicated whether the elements of the parent item should be marked as grouped.
   * @return {string} The rendered HTML string.
   */
  PrimeFaces.widget.SelectOneMenu.prototype.renderSelectItems = function(parentItem, isGrouped) {
    var $this = this;
    var content = "";
    isGrouped = isGrouped || false;

    var opts = parentItem.children("option, optgroup");
    var hasOptgroup = opts.filter("optgroup").length > 0;

    if (!hasOptgroup && !isGrouped) {
      content += '<ul role="group" class="ui-selectonemenu-list ui-widget-content ui-widget ui-corner-all ui-helper-reset">';
    }

    content += opts.map(function(index, element) {
      return $this.renderSelectItem(element, isGrouped);
    }).get().join('');

    if (!hasOptgroup && !isGrouped) {
      content += '</ul>';
    }

    return content;
  };

  /**
   * Renders panel HTML code for one select item (group).
   * @private
   * @param {JQuery} item An option (group) for which to render HTML code.
   * @param {boolean} [isGrouped] Indicates whether the item is part of a group.
   * @return {string} The rendered HTML string.
   */
  PrimeFaces.widget.SelectOneMenu.prototype.renderSelectItem = function(item, isGrouped) {
    var $item = $(item);
    var isOptgroup = item.tagName === "OPTGROUP";
    var id = PrimeFaces.uuid();
    var escape = $item.data("escape");
    var label = isOptgroup ? $item.attr("label") : $item.text();

    if (escape) {
      label = isOptgroup ? $("<div>").text(label).html() : $item.html();
      label = label === "&nbsp;" ? " " : label;
    }

    var cssClass = isOptgroup ? "ui-selectonemenu-item-group ui-corner-all" :
      "ui-selectonemenu-item ui-selectonemenu-list-item ui-corner-all" +
      (isGrouped ? " ui-selectonemenu-item-group-children" : "") +
      ($item.data("noselection-option") ? " ui-noselection-option" : "");

    var content = isOptgroup ? '<ul role="group" class="ui-selectonemenu-list ui-widget-content ui-widget ui-corner-all ui-helper-reset" aria-labelledby="' + id + '">' : '';

    content += '<li id="' + id + '" class="' + cssClass + '" tabindex="-1" role="' + (isOptgroup ? "presentation" : "option") + '"';

    var title = $item.data("title");
    if (title) {
      content += ' title="' + PrimeFaces.escapeHTML(title) + '"';
    }
    if ($item.is(':disabled')) {
      content += ' disabled';
    }

    var dataLabel = escape ? label.replaceAll('"', '&quot;') : PrimeFaces.escapeHTML(label, true);
    content += ' data-label="' + dataLabel + '">' + label + '</li>';

    if (isOptgroup) {
      content += this.renderSelectItems($item, true) + '</ul>';
    }

    return content;
  };
  
   /**
     * Highlights the next option after the currently highlighted option in the overlay panel.
     * @private
     * @param {JQuery.TriggeredEvent} event The event of the keypress.
     */
    PrimeFaces.widget.SelectOneMenu.prototype.highlightNext = function(event) {
        // Get the currently active item
        var activeItem = this.getActiveItem();
        
        // Check if the panel is hidden
        var isHidden = this.panel.is(':hidden');
        
        // Create a selector for the next valid option
        // If panel is hidden, select the first option, otherwise select the first visible option
        var selector = '[role="option"]:not(.ui-state-disabled)' + (isHidden ? ':first' : ':visible:first');
        
        // Try to find the next valid option after the active item
        var next = activeItem.nextAll(selector);

        if (next.length === 0) {
            var index = this.items.index(activeItem) + 1;
            if (index >= 0) {
                next = this.items.eq(index);
            }
        }
    
        if(event.altKey) {
            this.show();
        }
        else if(next.length === 1) {
            if(this.panel.is(':hidden')) {
                this.selectItem(next);
            }
            else {
                this.highlightItem(next);
                PrimeFaces.scrollInView(this.itemsWrapper, next);
            }
            this.changeAriaValue(next);
        }

        event.preventDefault();
    };

    /**
     * Highlights the previous option before the currently highlighted option in the overlay panel.
     * @private
     * @param {JQuery.TriggeredEvent} event The event of the keypress.
     */
    PrimeFaces.widget.SelectOneMenu.prototype.highlightPrev = function(event) {
        // Get the currently active item
        var activeItem = this.getActiveItem();
        
        // Check if the panel is hidden
        var isHidden = this.panel.is(':hidden');
        
        // Create a selector for the next valid option
        // If panel is hidden, select the first option, otherwise select the first visible option
        var selector = '[role="option"]:not(.ui-state-disabled)' + (isHidden ? ':first' : ':visible:first');
        
        // Try to find the previous valid option after the active item
        var prev = activeItem.prevAll(selector);
        
        // If no next option found, look for the first option in the next group (ul)
        if (prev.length === 0) {
            var index = this.items.index(activeItem) - 1;
            if (index >= 0) {
                prev = this.items.eq(index);
            }
        }

        if(prev.length === 1) {
            if(this.panel.is(':hidden')) {
                this.selectItem(prev);
            }
            else {
                this.highlightItem(prev);
                PrimeFaces.scrollInView(this.itemsWrapper, prev);
            }
            this.changeAriaValue(prev);
        }

        event.preventDefault();
    };

};
```
